### PR TITLE
Database-analysis versioning

### PR DESCRIPTION
**Pull request by Graves067**
## 📝 Description

> adds new tables for database keeping track of old versions of the project analysis
**Closes:** #351 

---

## 🔧 Type of Change

- [x] 🐛 Bug fix (non-breaking change that fixes an issue)
- [x] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📚 Documentation added/updated
- [x] ✅ Test added/updated
- [ ] ♻️ Refactoring
- [ ] ⚡ Performance improvement

---

## 🧪 Testing

> please rebuild the docker instance
Docker compose down -v
Docker compose build --no-cache
Docker compose up

> run 3 tests using
 python -m pytest test/test_db_versioning_functions.py    

- [x] Test A

---

## ✓ Checklist

- [x] 🤖 GenAI was used in generating the code and I have performed a self-review of my own code
- [x] 💬 I have commented my code where needed
- [ ] 📖 I have made corresponding changes to the documentation
- [x] ⚠️ My changes generate no new warnings
- [x] ✅ I have added tests that prove my fix is effective or that my feature works and tests are passing locally
- [ ] 🔗 Any dependent changes have been merged and published in downstream modules
- [ ] 📱 Any UI changes have been checked to work on desktop, tablet, and/or mobile

---

### CODE REVIEW

File: test/test_db_versioning_functions.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        latest = versions[0]
+
+        self.assertEqual(latest['version_number'], 2)
+        self.assertEqual(latest['filename'], "file.json")
+        self.assertTrue(latest['is_current'])
+
+        selected = self.store.retrieve_selected_version(row_id, 1)
+        self.assertEqual(selected['content'], {"data": 1})
+
+    def test_restore_previous_version(self):
```
Comment (Puneet-Maan) [inline test/test_db_versioning_functions.py]: this does not restore, just needs a name change then.

**Review by Puneet-Maan — CHANGES_REQUESTED** [review]
good work on this Cameron. I added a few inline comments for some very small stuff if you wanna address that.

and then just a couple of things I noticed:

1. we now have two different ways to save versions: the Python helper and the SQL stored procedures. The Python code already creates version rows but the stored procedures also try to save the current version into project_versions using the same version number which can cause a duplicate key or version already exists error if someone calls the procedure later.

2. in update(), we read current_version, then write the next version. If two updates happen at the same time, they can both read the same current_version and both try to create the same next version number. I think wee can fix that by doing the update inside a transaction and using a row lock, like SELECT ... FOR UPDATE.

File: test/test_db_versioning_functions.py
Comment (ssmith86) [inline test/test_db_versioning_functions.py]: coverage is overall good. if you want to make it more robust you could add some edge cases like missing id, invalid inputs etc., but not required.

All tests pass:

Image

**Comment by Puneet-Maan** [comment]
@Graves067 Im not sure if the first issue has been addressed yet. Please advice.

> good work on this Cameron. I added a few inline comments for some very small stuff if you wanna address that.
> 
> and then just a couple of things I noticed:
> 
> 1. we now have two different ways to save versions: the Python helper and the SQL stored procedures. The Python code already creates version rows but the stored procedures also try to save the current version into project_versions using the same version number which can cause a duplicate key or version already exists error if someone calls the procedure later.
> 2. in update(), we read current_version, then write the next version. If two updates happen at the same time, they can both read the same current_version and both try to create the same next version number. I think wee can fix that by doing the update inside a transaction and using a row lock, like SELECT ... FOR UPDATE.
